File: docs/sphinx/snippets/cpp/using/backends/dynamics_state_batching.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include "cudaq/algorithms/evolve.h"
#include "cudaq/algorithms/integrator.h"
#include "cudaq/operators.h"
#include <cudaq.h>

int main() {
  // [Begin State Batching]

  // Qubit Hamiltonian
  auto hamiltonian = 2 * M_PI * 0.1 * cudaq::spin_op::x(0);

  // A single qubit with dimension 2.
  cudaq::dimension_map dimensions = {{0, 2}};

  // Initial states in the `SIC-POVM` set:
  // https://en.wikipedia.org/wiki/SIC-POVM
  auto psi_1 =
      cudaq::state::from_data(std::vector<std::complex<double>>{1.0, 0.0});
  auto psi_2 = cudaq::state::from_data(std::vector<std::complex<double>>{
      1.0 / std::sqrt(3.0), std::sqrt(2.0 / 3.0)});
  auto psi_3 = cudaq::state::from_data(std::vector<std::complex<double>>{
      1.0 / std::sqrt(3.0),
      std::sqrt(2.0 / 3.0) *
          std::exp(std::complex<double>{0.0, 1.0} * 2.0 * M_PI / 3.0)});
  auto psi_4 = cudaq::state::from_data(std::vector<std::complex<double>>{
      1.0 / std::sqrt(3.0),
      std::sqrt(2.0 / 3.0) *
          std::exp(std::complex<double>{0.0, 1.0} * 4.0 * M_PI / 3.0)});
  // We run the evolution for all the SIC state to determine the process
  // tomography.
  std::vector<cudaq::state> sic_states = {psi_1, psi_2, psi_3, psi_4};

  // Schedule of time steps.
  std::vector<double> steps = cudaq::linspace(0.0, 10.0, 101);
  cudaq::schedule schedule(steps);

  // A default Runge-`Kutta` integrator
  cudaq::integrators::runge_kutta integrator;

  // Run the batch simulation.
  auto evolve_results = cudaq::evolve(
      hamiltonian, dimensions, schedule, sic_states, integrator, {},
      {cudaq::spin_op::x(0), cudaq::spin_op::y(0), cudaq::spin_op::z(0)},
      cudaq::IntermediateResultSave::ExpectationValue);

  // [End State Batching]

  return 0;
}

File: pr-4370/sphinx/snippets/python/using/backends/dynamics_state_batching.py
# ============================================================================ #
# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

#[Begin State Batching]

import cudaq
import cupy as cp
import numpy as np
from cudaq import spin, Schedule, RungeKuttaIntegrator
# Set the target to our dynamics simulator
cudaq.set_target("dynamics")

# Qubit Hamiltonian
hamiltonian = 2 * np.pi * 0.1 * spin.x(0)

# Dimensions of sub-system. We only have a single degree of freedom of dimension 2 (two-level system).
dimensions = {0: 2}

# Initial states in the `SIC-POVM` set: https://en.wikipedia.org/wiki/SIC-POVM
psi_1 = cudaq.State.from_data(cp.array([1.0, 0.0], dtype=cp.complex128))
psi_2 = cudaq.State.from_data(
    cp.array([1.0 / np.sqrt(3.0), np.sqrt(2.0 / 3.0)], dtype=cp.complex128))
psi_3 = cudaq.State.from_data(
    cp.array([
        1.0 / np.sqrt(3.0),
        np.sqrt(2.0 / 3.0) * np.exp(1j * 2.0 * np.pi / 3.0)
    ],
             dtype=cp.complex128))
psi_4 = cudaq.State.from_data(
    cp.array([
        1.0 / np.sqrt(3.0),
        np.sqrt(2.0 / 3.0) * np.exp(1j * 4.0 * np.pi / 3.0)
    ],
             dtype=cp.complex128))

# We run the evolution for all the SIC state to determine the process tomography.
sic_states = [psi_1, psi_2, psi_3, psi_4]
# Schedule of time steps.
steps = np.linspace(0, 10, 101)
schedule = Schedule(steps, ["time"])

# Run the batch simulation.
evolution_results = cudaq.evolve(
    hamiltonian,
    dimensions,
    schedule,
    sic_states,
    observables=[spin.x(0), spin.y(0), spin.z(0)],
    collapse_operators=[],
    store_intermediate_results=cudaq.IntermediateResultSave.EXPECTATION_VALUE,
    integrator=RungeKuttaIntegrator())

#[End State Batching]
